### PR TITLE
Try to fix shutdown test in edge cases

### LINKED ISSUE #85259: TestAutograd.test_thread_shutdown may fail sometimes
### 🐛 Describe the bug

When running this test e.g. via `python test_autograd.py TestAutograd.test_thread_shutdown` it may fail with the following sequence of events (i.e. this is a race condition):

- The test expects `"PYTORCH_API_USAGE torch.autograd.thread_shutdown"` at https://github.com/pytorch/pytorch/blob/d561aa944b7e777eb0575be2427e26a86df85f11/test/test_autograd.py#L4556
- This should be done by: https://github.com/pytorch/pytorch/blob/d561aa944b7e777eb0575be2427e26a86df85f11/torch/csrc/autograd/engine.cpp#L445
- That in turn is only done via a "ShutdownTask" which (on a successful test) will be added via `Engine::~Engine()` https://github.com/pytorch/pytorch/blob/d561aa944b7e777eb0575be2427e26a86df85f11/torch/csrc/autograd/engine.cpp#L259 calling `stop()` https://github.com/pytorch/pytorch/blob/d561aa944b7e777eb0575be2427e26a86df85f11/torch/csrc/autograd/engine.cpp#L283

However that ShutdownTask is only pushed when all queues are empty, see https://github.com/pytorch/pytorch/blob/d561aa944b7e777eb0575be2427e26a86df85f11/torch/csrc/autograd/engine.cpp#L279

As the [test code](https://github.com/pytorch/pytorch/blob/d561aa944b7e777eb0575be2427e26a86df85f11/test/test_autograd.py#L4536-L4550) is very simple it is possible that the the "run_backward" task pushed at the end was not yet picked up by the/a thread and hence is still in the/a queue when the Engine destructor is run during program shutdown. In that case the ShutdownTask will never be pushed and the test fails with

```
FAIL: test_thread_shutdown (__main__.TestAutograd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_autograd.py", line 4360, in test_thread_shutdown
    self.assertRegex(s, "PYTORCH_API_USAGE torch.autograd.thread_shutdown")
AssertionError: Regex didn't match: 'PYTORCH_API_USAGE torch.autograd.thread_shutdown' not found in 'PYTORCH_API_USAGE torch.python.import\nPYTORCH_API_USAGE c10d.python.import\nPYTORCH_API_USAGE tensor.create\n'
```

### Versions

Noticed with PyTorch 1.9.0 on an Intel Cascade Lake system with V100 GPUs: https://github.com/easybuilders/easybuild-easyconfigs/issues/16233

But may still apply as the relevant code is unchanged. See the referenced parts from current master above.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by albanD**
Fixes https://github.com/pytorch/pytorch/issues/85259
See the issue for debugging details.
tl;dr: when a worker thread is actually used, make sure it is initialized before exiting.
Yes, it is very unlikely it will take >10s to initialize but it is what seems to happen.

### CODE REVIEW

**Review by soulitzer — APPROVED** [review]
Cool!

**Review by ezyang — APPROVED** [review]
